Fix auth code cleanup function

bruce-ricard opened the pull request:
It will now properly only clean once every 5 minutes

bruce-ricard committed:
refactor: introduce clock

* so that we can mock it later in the tests

bruce-ricard committed:
fix: broken tests

* the tests shouln't expect that the codes are cleaned up right away,
  as they potentially get cleaned up only 5 minutes later. We added a
  mock clock to simulate the time going forward more than five minutes
  so that the tests have the correct expectations.

* this change will make even more sense with the next commit, which
  will fix a bug in that logic

bruce-ricard committed:
fix: bug in cleanup initialization

* the previous code was just adding 5 minutes to the last "last"
  cleanup, which is wrong. The "last" variable is meant to store the
  actual last cleanup time. This is what this changes does.

bruce-ricard committed:
refactor: make final

bruce-ricard committed:
refactor: simplify

* cleanup suggested by IntelliJ. I wouldn't have realized that this
  test was always "true".

bruce-ricard committed:
refactor: rename

bruce-ricard committed:
refactor: rename

bruce-ricard commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -450,5 +468,24 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
         }
     }
 
+    public class TestClock extends Clock {
+        public Queue<Instant> instants = new LinkedList<Instant>();
+
+        @Override
+        public ZoneId getZone() {
+            return ZoneId.systemDefault();
+        }
+
+        @Override
+        public Clock withZone(ZoneId zone) {
+            return this;
+        }
+
+        @Override
+        public Instant instant() {
+            return instants.remove();
+        }
+    }
+
```
That's the only way I found to do what I wanted to do. Please advice if you know a better way.

strehle replied:
my advise is , use mockito for the expiration, see https://github.com/cloudfoundry/uaa/pull/2292

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -450,5 +468,24 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
         }
     }
 
+    public class TestClock extends Clock {
```
The class name is a bit confusing - sounds like this is a unit test class.

bruce-ricard replied:
OK, interesting. IMO `ClockTest` is a test class. `TestClock` is a clock for tests.

But actually I think that I can improve its name, so I'll try that.

swalchemist replied:
If I write a unit test for the Clock class, I would name it ClockTest, which sounds like TestClock. So "Test" tends to be overused as a name.

hsinn0 replied:
> ClockTest is a test class. TestClock is a clock for tests

Yep, that conforms with the Java test class naming pattern.

bruce-ricard replied:
I renamed the class to something more specific.

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -206,32 +210,46 @@ void retrieveNonExistentToken() {
             legacyCodeServices.createAuthorizationCode(clientAuthentication);
         }
         assertThat(jdbcTemplate.queryForObject("SELECT count(*) FROM oauth_code", Integer.class), is(count));
         jdbcTemplate.update("UPDATE oauth_code SET created = ?", new Timestamp(System.currentTimeMillis() - (2 * oneday)));
```
If we're mocking the clock for this test, it would be best to choose a fixed point in time instead of starting it based on the system clock. Otherwise, you'll get clock ticks that you can't control for.

bruce-ricard replied:
It won't tick though. I'm creating a static clock that will return that exact time.

I don't think that I understand what you mean.

swalchemist replied:
I was referring to things like new `Timestamp(System.currentTimeMillis() - (2 * oneday))` that are still using the system clock.

bruce-ricard replied:
Right. That's old code that writes stuff to the DB in the middle of the test.

I would have never written this like that in the first place, but again I don't think that it's that important to fix this now. Another commit/PR can do that.

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -206,32 +210,46 @@ void retrieveNonExistentToken() {
         assertThat(jdbcTemplate.queryForObject("SELECT count(*) FROM oauth_code", Integer.class), is(count));
         jdbcTemplate.update("UPDATE oauth_code SET created = ?", new Timestamp(System.currentTimeMillis() - (2 * oneday)));
+        clock.instants.add(Instant.now());
         assertThrows(InvalidGrantException.class, () -> store.consumeAuthorizationCode("non-existent"));
```
The three blocks of assertions make this test a bit hard to follow. Maybe it wants to be two or three different tests?

bruce-ricard replied:
Same as the other comment about code I didn't change.

swalchemist replied:
You did make the code blocks longer, though. An easy fix might be adding some whitespace.

bruce-ricard replied:
I like new lines. Let me know what you think of this.

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -206,32 +210,46 @@ void retrieveNonExistentToken() {
     }
 
     @Test
     void cleanUpLegacyCodesCodesWithoutExpiresAtAfter3Days() {
```
This test method name is unparseable. Could it be reduced to English?

bruce-ricard replied:
I don't disagree. I hadn't read it until now.

It's a bit awkward to read a comment about an old line that I didn't touch. Are you asking me to fix that in this PR?

swalchemist replied:
Yes, though I wouldn't insist on it since you didn't introduce that name in this PR.

bruce-ricard replied:
OK the test was indeed nasty. I refactored it, and renamed the test.

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -450,5 +468,24 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
     }
 
+    public class TestClock extends Clock {
+        public Queue<Instant> instants = new LinkedList<Instant>();
```
Could you improve the name here to make it more clear that this is a list of instances that will be returned by the class?

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -450,5 +468,24 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
+        }
+
+        @Override
+        public Instant instant() {
```
Safe to assume that this is called when now() is invoked? Why not modify now() directly?

bruce-ricard replied:
`now()` is not a `Clock` method, it's an `Instant` method, that calls the `Clock#instant()` method.

swalchemist commented on server/src/test/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStoreTests.java:
```diff
@@ -450,5 +468,24 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
+
+        @Override
+        public ZoneId getZone() {
+            return ZoneId.systemDefault();
```
Using the system default zone wouldn't be good in some situations, but probably works fine for this test.

swalchemist commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStore.java:
```diff
@@ -83,24 +84,36 @@ public class UaaTokenStore implements AuthorizationCodeServices {
-        int tries = 0;
-        while ((tries++)<=max_tries) {
+        int attempt = 0;
+        while (true) {
```
The check on maxTries has moved inside a catch block. Is there a potential infinite loop here?

bruce-ricard replied:
It hasn't moved. It disappeared. As I explain in the commit message, this expression is always `true`, as explained by IntelliJ.

hsinn0 replied:
~~Although the change is not going to cause an infinite loop because of the `throw` in the `catch` block, this change does modify the code behavior.
The old code will gracefully exit the while loop when `tries` becomes 3.  The new code will throw `DataIntegrityViolationException`.   Notice that the post-increment operator (`++`) was used in the condition in the old code.~~
No that is not it.  Let me re-read this.  I see.  Part of the method is not well-written.  So the logic is:
- return immediately if `template.update()` was successful but try couple of more times in if the update was not successful.

The throw and rethrow of `DataIntegrityViolationException` is unnecessary.  The code can just count the number of failed updates then throw exception once if the `max_tries` go over the `tries`.  This code should be changed for code efficiency and performance reason.

bruce-ricard replied:
>Part of the method is not well-written

What was your first clue?

bruce-ricard replied:
>The throw and rethrow of DataIntegrityViolationException is unnecessary.


I agree, this is pretty crazy.

Are you saying that you want this changed as part of this PR?

hsinn0 replied:
> > Part of the method is not well-written
> 
> What was your first clue?

I could only see what the old code was doing after expanding the hidden code block in this PR as that code was not part of this PR change.

hsinn0 replied:
> > The throw and rethrow of DataIntegrityViolationException is unnecessary.
> 
> I agree, this is pretty crazy.
> 
> Are you saying that you want this changed as part of this PR?

It can be a separate PR if you prefer that way.

swalchemist replied:
Bruce is right, this was already in the catch block.

But I do not see how `(tries++)<=max_tries` is always true, and the refactor that my IntelliJ suggests that the logic move inside the loop, not that it be removed entirely. I'm still concerned about a potential infinite loop, and I'm curious why there isn't a test that catches it.

bruce-ricard replied:
I'm gonna makes more PRs on this code anyways.

bruce-ricard replied:
> the refactor that my IntelliJ suggests that the logic move inside the loop, not that it be removed entirely.

That's pretty amazing that our IntelliJs suggest 2 different refactors.

bruce-ricard replied:
I reverted that change. I'll make another short PR for it.

hsinn0 commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStore.java:
```diff
@@ -83,24 +84,36 @@ public class UaaTokenStore implements AuthorizationCodeServices {
     }
 
     public UaaTokenStore(DataSource dataSource, Duration expirationTime) {
+        this(dataSource, expirationTime, Clock.systemDefaultZone());
```
Methods such as `Instant.now()` returns systemUTC instant.  To avoid potential timezone mix-up, we'd better use `Clock.systemUTC()` here.

bruce-ricard replied:
That's a good point. I don't think that I understand very much the difference between the two.

As far as I can tell, `Instant.now()` will return the same value, regardless of the clock. But then it may get mixed up if we ask for the date of the instant? 

I'm gonna keep reading a bit on it.

bruce-ricard replied:
I'm super confused.

To convert an instant to a date, you are forced to pass the zone again, which makes complete sense. That I understand.
```
Instant.now().atZone(ZoneId.systemDefault()).getDayOfMonth();
```

But then why is there a `Clock#getZone()` method. I quickly took a look at the library code, the zone doesn't appear to be used for anything in `Instant`. I'm wondering if we can just pass `null` here, as we are not going to use the Zone.

bruce-ricard replied:
I meant passing `null` for the ZoneId.

bruce-ricard replied:
Alright my quick investigation found that `Instant` doesn't use the Clock's zone for anything that we care about, so any System Clock will result in the same behavior.

I would like to do `Clock.system(null)` but there's a null check on the zone parameter, so we can't do that.

With all that information, I lean towards just leaving the code the way it is. Let me know what you think.

hsinn0 replied:
Java 11 Instant.now() source code:
>     public static Instant now() {
>         return Clock.systemUTC().instant();
>     }
>

bruce-ricard replied:
Which is unbelievably confusing, because the `.instant()` call doesn't use the Clock's zone.

bruce-ricard committed:
refactor: test clock

* I changed my mind with regards to the test clock. It shouldn't have
  such fine grain setting of the time, through a queue, as this is a
  sign that the test is too white box. Like all injected dependencies,
  the clock needs to be configured to do one thing, and do it until
  it's configured to do something else.

strehle commented:
Hi,
can you please at least have a short look at https://github.com/cloudfoundry/uaa/pull/2292 because to me this fix is small enough and we dont need to manipulate the clock with mocking the expiration

strehle reviewed (changes requested):
please check
https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2291
Missing conditions in code coverage

bruce-ricard committed:
formatting: new lines

* split blocks to try to clarify which ones are test setup and which
  ones run assertions

bruce-ricard committed:
refactor: test

* use the test clock to simulate going forward in time, rather than
  hacking the creation date in the DB with a SQL statement.

bruce-ricard commented:
> please check https://sonarcloud.io/summary/new_code?id=cloudfoundry-identity-parent&pullRequest=2291 Missing conditions in code coverage

This is a false positive. I didn't add any true new code. It's just complaining because it doesn't realize I refactored code.

I'll try to see if I can add coverage regardless.

strehle commented:
> I'll try to see if I can add coverage regardless.

yes and it shows that there is new complexity with the while(true) and exit if

`
} catch (DataIntegrityViolationException exists) {
                if (attempt>=maxAttempts) throw exists;
            }
`

So is this change really needed?

strehle commented on server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStore.java:
```diff
@@ -213,10 +225,10 @@ else if (map.get(USER_AUTHENTICATION_UAA_PRINCIPAL)!=null) {
         if (enoughTimeHasPassedSinceLastExpirationClean(last, now)) {
             //avoid concurrent deletes from the same UAA - performance improvement
-            if (lastClean.compareAndSet(last, last.plus(getExpirationTime()))) {
+            if (lastClean.compareAndSet(last, now)) {
```
to me this is the only line which need to be fixed, so compareAndSet needs the correct new value which is now

hsinn0 replied:
> to me this is the only line which need to be fixed

I agree, now especially as the other alternate PR does not have sleep() in the unit test code anymore.  My preference now is going with https://github.com/cloudfoundry/uaa/pull/2292.

strehle commented:
only mention this for completeness : Another way to fix this would be to remove the internal cleanup and the setup of a 
task scheduler, e.g. https://www.concretepage.com/spring/spring-task-scheduler in order to regular cleanup this (and maybe the other, e.g. revocable_token) table(s) . We have this cleanup in our fork

bruce-ricard commented:
> only mention this for completeness : Another way to fix this would be to remove the internal cleanup and the setup of a task scheduler, e.g. https://www.concretepage.com/spring/spring-task-scheduler in order to regular cleanup this (and maybe the other, e.g. revocable_token) table(s) . We have this cleanup in our fork

I agree with that. I don't like the way the cleanup is currently happening. That would be a much larger code change though. If you think that you can easily bring it from your fork, that would be great.

bruce-ricard committed:
Revert "refactor: simplify"

This reverts commit f5ba7e688bdd08a77bd7c4535c0f54c3edefd74a.

Reverting this commit for now, to take it out of the PR. We will make
another PR for this.

strehle commented:
I agree that we can / should refactor this class, but to get a fixed release, can we agree on a minimal fix ?
https://github.com/cloudfoundry/uaa/pull/2311 because I think there we are all the same opinion ?

strehle commented:
@bruce-ricard I checked JdbcRevocableTokenProvisioning or JdbcExpiringCodeStore, and then realized, that Filip has introduced a time interface for such situations already and since I would stay with UAA currenty "style" I would prefere to have the TimeService as layer to abstract the time and not introducing a new variant with Clock

See
https://github.com/cloudfoundry/uaa/blob/develop/uaa/src/main/webapp/WEB-INF/spring/oauth-endpoints.xml#L535
or
https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/codestore/JdbcExpiringCodeStore.java#L167-L174

So I vote for replacing clock with timeservice and then of course we can test similar situations because we can mock the time, e.g. https://github.com/cloudfoundry/uaa/blob/develop/server/src/test/java/org/cloudfoundry/identity/uaa/oauth/TokenValidityResolverTest.java#L24-L26

swalchemist commented:
Are we still going to use the change in this PR?

hsinn0 commented:
> Are we still going to use the change in this PR?

This PR should be closed.  For the immediate fix, we already merged #2292.  For the follow-up refactoring, we have #2315. This was discussed during OSS sync-up meeting on 5/3/2023.

Tallicia commented:
I'm closing this per discussion above. Please, if it needs to be closed, explain why and close it.
